cleverhans/experimental/certification/optimization.py

```diff
@@ -52,6 +56,51 @@ def __init__(self, dual_formulation_object, sess, optimization_params):
     # Create graph for optimization
     self.prepare_for_optimization()
 
+  def project_dual(self):
```

Comment:
This function is essentially does not depend on anything declared in this class except for dual_object, and it creates another version of dual object. So it looks like DualFormulation would be a better place for this function.
In such case you slighlty shorten lines in this function (because you type `self` instead of `self.dual_object`), also you don't need to add `min_eig_val_h` (which is not dual variable) into `dual_var` dictionary:

```
class DualFormulation(object):
  
    def make_projected_dual(self):
      projected_nu = ...
      min_eig_h = ...
      projected_lambda_pos = [tf.identity(x) for x in self.lambda_pos]
      ...
      projected_dual_object = DualFormulation(...)
      projected_dual_object.min_eig_val_h = min_eig_h
      return projected_dual_object

```